overlord/devicestate/users.go

```diff
@@ -0,0 +1,359 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package devicestate
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os/user"
+	"path/filepath"
+	"time"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/overlord/assertstate"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/strutil"
+)
+
+type UserResponse struct {
+	Username string
+	SSHKeys  []string
+}
+
+var (
+	osutilAddUser = osutil.AddUser
+	osutilDelUser = osutil.DelUser
+	userLookup    = user.Lookup
+)
+
+type UserError struct {
+	Err      error
+	Internal bool
+}
+
+func (e UserError) Error() error {
+	return e.Err
+}
+
+func (e UserError) IsInternal() bool {
+	return e.Internal
+}
+
+func RemoveUser(st *state.State, username string) (*auth.UserState, *UserError) {
+	// TODO: allow to remove user entries by email as well
+
+	// catch silly errors
+	if username == "" {
+		return nil, &UserError{Internal: false, Err: fmt.Errorf("need a username to remove")}
+	}
+	// check the user is known to snapd
+	st.Lock()
+	_, err := auth.UserByUsername(st, username)
+	st.Unlock()
+	if err == auth.ErrInvalidUser {
+		return nil, &UserError{Internal: false, Err: fmt.Errorf("user %q is not known", username)}
+	}
+	if err != nil {
+		return nil, &UserError{Internal: true, Err: err}
+	}
+
+	// first remove the system user
+	if err := osutilDelUser(username, &osutil.DelUserOptions{ExtraUsers: !release.OnClassic}); err != nil {
+		return nil, &UserError{Internal: true, Err: err}
+	}
+
+	// then the UserState
+	st.Lock()
+	u, err := auth.RemoveUserByUsername(st, username)
+	st.Unlock()
+	// ErrInvalidUser means "not found" in this case
+	if err != nil && err != auth.ErrInvalidUser {
+		return nil, &UserError{Internal: true, Err: err}
+	}
+	return u, nil
+}
+
+func CreateUser(st *state.State, mgr *DeviceManager, sudoer bool, createKnown bool, email string) ([]UserResponse, *UserError) {
```

Comment:
I'd suggest splitting this into `CreateUser` and `CreateKnownUser`; if needed, you can make an internal function to share their common parts of the code, but from the caller point of view these two functionality seem different enough to deserve a different public function.

There's also the question of the `sudoer` and the `createKnown` flags, which I suspect are incompatible (at least, when the email is not given, I see that the `sudoer` flag is ignored); can one specify the `sudoer` flag to `true` when creating a known user? I think that splitting the two functions can make things more clear.

Reply:
This is now split into two functions, though the second function is `CreateKnownUsers`, as it can create more than one user.
Splitting made is cleaner, at least `known` division is now cleaner as for `CreateUser` we always need to provide email, while for `CreateKnownUsers` email is optional. And we do not need always manager either.
I'm reluctant to remove `sudoer` flag as this was all (including `known` flag) part of the original implementation. So in case we are missing something I'd keep sudoer flag for now.